pkg/controller/shared_state.go

```diff
@@ -555,8 +555,97 @@ func (c *SharedState) addVolumeAndClaim(azVolumeName, pvName, pvClaimName string
 	c.claimToVolumeMap.Store(pvClaimName, azVolumeName)
 }
 
-func (c *SharedState) deletePV(pvName string) {
+func (c *SharedState) deletePV(pvName string) error {
+	var err error
+
+	ctx := context.Background()
+	ctx, w := workflow.New(ctx, workflow.WithDetails())
+	defer func() { w.Finish(err) }()
+
 	c.pvToVolumeMap.Delete(pvName)
+
+	// fetch AzVolume CRI for the PV, and delete if it's a preprovisioned volume
+	var azVolumeList azdiskv1beta2.AzVolumeList
+	req, err := azureutils.CreateLabelRequirements(consts.PvNameLabel, selection.Equals, pvName)
+	if err != nil {
+		return err
+	}
+	err = c.cachedClient.List(ctx, &azVolumeList, &client.ListOptions{LabelSelector: labels.NewSelector().Add(*req)})
+	if err != nil && !apiErrors.IsNotFound(err) {
+		return err
+	} else if apiErrors.IsNotFound(err) || len(azVolumeList.Items) == 0 {
+		return nil
+	}
+
+	// deletion timestamp is set and AzVolume reconcliler will handle the delete request.
+	// The volume itself will not be deleted.
+	azVolume := azVolumeList.Items[0]
+	w.AddDetailToLogger(workflow.GetObjectDetails(&azVolume)...)
+
+	if azureutils.MapContains(azVolume.Status.Annotations, consts.PreProvisionedVolumeAnnotation) {
+		return nil
+	}
+
+	// AzVolume does exist and needs to be deleted
+	// add annotation to mark AzVolumeAttachment cleanup
+	err = c.cachedClient.Delete(ctx, &azVolume)
```

Comment:
Shouldn't we add the cleanup annotation before deleting the `AzVolume`? If the `azvolume` controller sees the delete  *before* the annotation is added, won't it try to delete the pre-provisioned disk?

Reply:
The azVolume controller would see that an `AzVolume` deletion was requested without a `volumeDetachRequestAnnotation`, in which case it would consider it a uninstall scenario and detach replicas without deleting the volume.

But good catch. The annotations need to be added before issuing a delete request. I was confusing upcoming Huichan's change for removing the cleanup annotation for pre-provisioned volumes.